src/pages/resultsView/plots/PlotsTabUtils.tsx

```diff
@@ -414,39 +427,120 @@ function makeAxisDataPromise_Clinical(
 function makeAxisDataPromise_Molecular(
     entrezGeneId:number,
     molecularProfileId:string,
+    mutationCache:MobxPromiseCache<{entrezGeneId:number}, Mutation[]>,
     numericGeneMolecularDataCache:MobxPromiseCache<{entrezGeneId:number, molecularProfileId:string}, NumericGeneMolecularData[]>,
     entrezGeneIdToGene:MobxPromise<{[entrezGeneId:number]:Gene}>,
-    molecularProfileIdToMolecularProfile:MobxPromise<{[molecularProfileId:string]:MolecularProfile}>
+    molecularProfileIdToMolecularProfile:MobxPromise<{[molecularProfileId:string]:MolecularProfile}>,
+    mutationCountBy: MutationCountBy,
+    coverageInformation:MobxPromise<CoverageInformation>,
+    samples:MobxPromise<Sample[]>
 ):MobxPromise<IAxisData> {
-    const promise = numericGeneMolecularDataCache.get({
-        entrezGeneId: entrezGeneId,
-        molecularProfileId: molecularProfileId
-    });
+    let promise:MobxPromise<any>;/* = ;*/
     return remoteData({
-        await:()=>[promise, entrezGeneIdToGene, molecularProfileIdToMolecularProfile],
+        await:()=>{
+            const ret:MobxPromise<any>[] = [];
+            if (molecularProfileIdToMolecularProfile.isComplete) {
+                if (molecularProfileIdToMolecularProfile.result![molecularProfileId].molecularAlterationType ===
+                        AlterationTypeConstants.MUTATION_EXTENDED) {
+                    // mutation profile
+                    promise = mutationCache.get({ entrezGeneId });
+                    ret.push(coverageInformation);
+                    ret.push(samples);
+                } else {
+                    // non-mutation profile
+                    promise = numericGeneMolecularDataCache.get({ entrezGeneId, molecularProfileId });
+                }
+                ret.push(promise);
+            } else {
+                ret.push(molecularProfileIdToMolecularProfile);
+            }
+            ret.push(entrezGeneIdToGene);
+            return ret;
+        },
         invoke:()=>{
             const profile = molecularProfileIdToMolecularProfile.result![molecularProfileId];
-            const isDiscreteCna = (profile.molecularAlterationType === AlterationTypeConstants.COPY_NUMBER_ALTERATION
-                                    && profile.datatype === "DISCRETE");
-            const data:NumericGeneMolecularData[] = promise.result!;
-            return Promise.resolve({
-                data: data.map(d=>{
-                    let value = d.value;
-                    if (isDiscreteCna) {
-                        const appearance = (cnaToAppearance as any)[d.value];
-                        if (appearance) {
-                            value = appearance.legendLabel;
+            const hugoGeneSymbol = entrezGeneIdToGene.result![entrezGeneId].hugoGeneSymbol;
```

Comment:
seems like a lot of logic here.  can we break it down/externalize and maybe test?

Reply:
sure, that makes sense